### PR TITLE
Problem: Missing ICA auth module for enabling interchain accounts

### PR DESCRIPTION
**Pull request by devashishdxt**
Solution: Added a basic ICA auth module

As described in #719, interchain accounts feature will be divided in three separate PRs. This is the first one which adds a basic ICA auth module.

Ref: https://github.com/cosmos/ibc-go/blob/v3.0.0/docs/apps/interchain-accounts/auth-modules.md

### CODE REVIEW

**Review by tomtau — COMMENTED** [review]
does it need to be a separate module or could the functionality be added to an existing module? I'm thinking that x/chainmain is pretty much empty, but ok to keep it separate too

**Comment by devashishdxt** [comment]
> does it need to be a separate module or could the functionality be added to an existing module? I'm thinking that x/chainmain is pretty much empty, but ok to keep it separate too

It does not need to be in a separate module but it's better to keep it separate for clarity and logical separation.

**Comment by tomtau** [comment]
ok, I'll need to read https://github.com/cosmos/ibc-go/blob/v3.0.0/docs/apps/interchain-accounts/auth-modules.md first

**Review by tomtau — COMMENTED** [review]
I'm a bit puzzled about the `OnAcknowledgementPacket` section -- is that done in ibc-go?

File: x/icaauth/module_ibc.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+	order channeltypes.Order,
+	connectionHops []string,
+	portID,
+	channelID string,
+	channelCap *capabilitytypes.Capability,
+	counterparty channeltypes.Counterparty,
+	counterpartyVersion string,
+) (string, error) {
+	// https://github.com/cosmos/ibc-go/blob/v3.0.0/docs/apps/interchain-accounts/auth-modules.md#ibcmodule-implementation
+	panic("UNIMPLEMENTED")
```
[the remark this answers is not in the document]
Comment (devashishdxt) [reply x/icaauth/module_ibc.go:60]: These will never be called. So, it doesn't really matter if we panic or return an error. I'll change these to error just to be safe.

**Comment by devashishdxt** [comment]
> I'm a bit puzzled about the `OnAcknowledgementPacket` section -- is that done in ibc-go?

Yes. For packet acknowledgement, both `ica-auth` and `ica-controller` (which is in `ibc-go`) will be called. Ref: https://github.com/cosmos/ibc-go/blob/f83aa8201eaa7b9daa2b34e4e44a219066e6e958/modules/apps/27-interchain-accounts/controller/ibc_module.go#L136

File: x/icaauth/module_ibc.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (devashishdxt) [reply x/icaauth/module_ibc.go]: Done

**Review by tomtau — APPROVED** [review]
it's probably ok -- the app wiring, upgrade handler etc. will be a separate PR if I understand the plan

File: x/icaauth/module.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+	return []abci.ValidatorUpdate{}
+}
+
+// ExportGenesis returns the capability module's exported genesis state as raw JSON bytes.
+func (am AppModule) ExportGenesis(ctx sdk.Context, cdc codec.JSONCodec) json.RawMessage {
+	return nil
+}
+
+// ConsensusVersion implements ConsensusVersion.
+func (AppModule) ConsensusVersion() uint64 { return 2 }
```
[the remark this answers is not in the document]
Comment (devashishdxt) [reply x/icaauth/module.go:143]: I'll change it to 1. It was auto-generated.

File: x/icaauth/types/message_submit_tx.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+	return msgs, nil
+}
+
+func (msg MsgSubmitTx) CalculateTimeoutDuration() time.Duration {
+	var timeoutDuration time.Duration
+
+	if msg.TimeoutDuration != nil {
+		timeoutDuration = *msg.TimeoutDuration
+	} else {
+		timeoutDuration = time.Minute * 5
```
[the remark this answers is not in the document]
Comment (devashishdxt) [reply x/icaauth/types/message_submit_tx.go:56]: ICA uses ordered channels and the channel closes if any packet times out. 5 minutes seems like a safe timeout duration to me. We can change it in future if we want.

[the remark this answers is not in the document]
Comment (devashishdxt) [reply x/icaauth/types/message_submit_tx.go:56]: Or we can just return error here and make it necessary for user to provide?

File: x/icaauth/module.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (devashishdxt) [reply x/icaauth/module.go]: Done

File: x/icaauth/types/message_submit_tx.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (devashishdxt) [reply x/icaauth/types/message_submit_tx.go:56]: Merging this PR for now. Created an issue to discuss this further: https://github.com/crypto-org-chain/chain-main/issues/731

**Comment by tomtau** [comment]
@devashishdxt would a different protobuf generator use non-deprecated packages?

**Comment by devashishdxt** [comment]
> @devashishdxt would a different protobuf generator version use non-deprecated packages?

Maybe if we upgrade to newer version then it won't use deprecated packages. But, I don't think we can stop using these deprecated packages in our code until `cosmos-sdk` removes/upgrades it in their interfaces.

**Comment by devashishdxt** [comment]
> @devashishdxt would a different protobuf generator version use non-deprecated packages?

I can see that `cosmos-sdk` is using these packages even in `v0.46.0-alpha3`.

**Pull request merged**